File: backend/app/schemas/poster.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field, field_validator

from app.core.categories import normalize


class PosterCreateIn(BaseModel):
    """Заводит карточку афиши. Доступно только оператору — это не пользовательский контент."""

    title: str = Field(..., min_length=2, max_length=200)
    description: str | None = None
    category: str | None = Field(default=None, max_length=60)
    starts_at: datetime
    ends_at: datetime | None = None
    venue: str | None = Field(default=None, max_length=200)
    address: str | None = None
    latitude: float = Field(..., ge=-90, le=90)
    longitude: float = Field(..., ge=-180, le=180)
    price_from: float | None = Field(default=None, ge=0)
    is_free: bool = False
    image_url: str | None = None
    source_url: str | None = None
    source_name: str | None = Field(default=None, max_length=120)

    @field_validator("category")
    @classmethod
    def _norm(cls, v: str | None) -> str | None:
        return normalize(v)


class PosterUpdateIn(BaseModel):
    title: str | None = Field(default=None, min_length=2, max_length=200)
    description: str | None = None
    category: str | None = Field(default=None, max_length=60)
    starts_at: datetime | None = None
    ends_at: datetime | None = None
    venue: str | None = Field(default=None, max_length=200)
    address: str | None = None
    price_from: float | None = Field(default=None, ge=0)
    is_free: bool | None = None
    image_url: str | None = None
    source_url: str | None = None
    source_name: str | None = Field(default=None, max_length=120)
    status: str | None = Field(default=None, pattern="^(published|hidden)$")

    @field_validator("category")
    @classmethod
    def _norm(cls, v: str | None) -> str | None:
        return normalize(v)


class PosterItem(BaseModel):
    id: uuid.UUID
    title: str
    description: str | None
    category: str | None
    starts_at: datetime
    ends_at: datetime | None
    venue: str | None
    address: str | None
    latitude: float
    longitude: float
    distance_km: float | None
    price_from: float | None
    is_free: bool
    image_url: str | None
    source_url: str | None
    source_name: str | None
    # Сколько компаний уже собирается на это мероприятие — главный мост к ядру продукта.
    gatherings_count: int
    status: str


class PosterOut(BaseModel):
    items: list[PosterItem]
    next_cursor: str | None


class CategoryItem(BaseModel):
    key: str
    title: str
    is_canonical: bool
    usage: int


class CategoriesOut(BaseModel):
    items: list[CategoryItem]

File: frontend/Skhodka/Skhodka/Models/Poster.swift
import Foundation

/// Афиша — чужое мероприятие, на которое можно сходить.
///
/// Отличается от события тем, что тут никто не собирает группу: концерт состоится
/// независимо от откликов, организатора среди пользователей нет, заявок и чата нет.
/// Связь с ядром одна: из карточки можно собрать компанию — тогда появится обычное
/// событие, ссылающееся сюда.
struct PosterItem: Decodable, Identifiable, Hashable {
    let id: String
    let title: String
    let description: String?
    let category: String?
    let startsAt: String
    let endsAt: String?
    let venue: String?
    let address: String?
    let latitude: Double
    let longitude: Double
    let distanceKm: Double?
    let priceFrom: Double?
    let isFree: Bool
    let imageURL: String?
    let sourceURL: String?
    let sourceName: String?
    /// Сколько компаний уже собирается сюда.
    let gatheringsCount: Int
    let status: String

    private enum CodingKeys: String, CodingKey {
        case id, title, description, category, venue, address, latitude, longitude, status
        case startsAt = "starts_at"
        case endsAt = "ends_at"
        case distanceKm = "distance_km"
        case priceFrom = "price_from"
        case isFree = "is_free"
        case imageURL = "image_url"
        case sourceURL = "source_url"
        case sourceName = "source_name"
        case gatheringsCount = "gatherings_count"
    }
}

struct PosterResponse: Decodable {
    let items: [PosterItem]
    let nextCursor: String?
    private enum CodingKeys: String, CodingKey {
        case items
        case nextCursor = "next_cursor"
    }
}

/// Категория из справочника: канонические плюс те, что вписали пользователи.
struct CategoryItem: Decodable, Hashable {
    let key: String
    let title: String
    let isCanonical: Bool
    let usage: Int

    private enum CodingKeys: String, CodingKey {
        case key, title, usage
        case isCanonical = "is_canonical"
    }
}

struct CategoriesResponse: Decodable {
    let items: [CategoryItem]
}
